keps/sig-api-machinery/365-paginated-lists/README.MD

```diff
@@ -152,6 +248,86 @@ malicious users cannot trick the server into serving results outside of its rang
 The storage layer must perform defensive checking on the provided value, check for path traversal attacks, and have 
 stable versioning for the continue token.
 
+#### watchcache
+
+However, not all LIST requests are served from etcd. Our API allows clients to opt-in for the LIST result to be
+served from kube-apiserver cache (called watchcache), being propagated via watch from etcd3.
+However, up until release 1.24, support for pagination wasn't added to watchcache. This makes the logic that
+we use extremely non-intuitive, as it effectively works as following:
+
+- if the `continuation` is set - always pass the request to etcd (and honor the LIMIT)
+- if resourveVersion='0' - ignore the LIMIT parameter and return result from watchcache
+- if resourceVersion=<some concrete RV> and LIMIT is set - pass the request to etcd (and honor the LIMIT)
+- if resourceVersion=<some concrete RV> and LIMIT is not set - serve the request from watchcache
+   - which effectively results in different semantics about data freshness depending on LIMIT being set or not
+
+Fixing that misleading experience is one of the issues that are GA blocker for the whole pagination feature.
+
+We propose to add support for pagination in watchcache as following:
+
+1. As of now, watchache (in addition to transaction log which is not relevant for this proposal) contains
+  the current (consistent at a given RV) state of all objects of a given type. This is currently stored
+  in client-go `Indexer`, which internally stores that in hash-map.
+  We propose to replace the usage of `Indexer` with an implementation that will be build on top of
+  BTree that supports copy-on-write semantic. We are planning to use the implementation from
+  [https://github.com/google/btree] (although it may require adding more extensive test coverage).
```

Comment:
at this point we will now have incrementally implemented 90% of etcd, vs 50%.

So the first meta question I want to discuss is whether the final shape is our idealized version of storage (a reference in memory version), and if so whether our approach is building towards something we should formally define.  Second is whether the etcd proxy / core etcd data structures does a better job of this than the approach we are taking.  Third will be identifying the core data structure tradeoff - is the btree the ideal structure for us vs other types (why you started here), or what key semantics are “must support”, so that we can decide whether the data structure is as minimal as possible.

Reply:
This is a good point but it makes me wonder if we'd just be better off moving etcd completely in-tree and running it directly out of the apiserver process.

Reply:
> at this point we will now have incrementally implemented 90% of etcd, vs 50%.

I don't agree with this one, but I also think it's not that important.

> So the first meta question I want to discuss is whether the final shape is our idealized version of storage (a reference in memory version), and if so whether our approach is building towards something we should formally define

The way I'm seeing this is that this is moving us towards the where I would like to land longer term. There is more one thing that I would like to happen - I described that in #108003 but those other changes are (a) not related to pagination (b) will be fairly small from watchcache perspective.

So answering the second part of the question - yes, I think defining this me explicitly makes perfect sense to - I would be happy to do that, but if possible I would prefer not to block this PR on it.

> Second is whether the etcd proxy / core etcd data structures does a better job of this than the approach we are taking.

etcd underneath is also using btree:
https://github.com/etcd-io/etcd/blob/7c472a964e870180a698836923a5325cd70e36bd/server/storage/mvcc/index.go#L21
[and actually the same btree that I'm proposing to use below]

So wrt the proposed change - we will be doing more-or-less how etcd is indexing data.
Now talking about watchcache as a whole, the watchable-store etcd implemented is probably better designed. And there is probably room for unification. I would be happy to address that in the "more formal definition" document that you suggested above, but such unification is orthogonal to the change proposed here.

> Third will be identifying the core data structure tradeoff - is the btree the ideal structure for us vs other types (why you started here), or what key semantics are “must support”, so that we can decide whether the data structure is as minimal as possible.

BTree is the standard answer that those kind of answers in the world. The fact that etcd is also using it internally (as I linked above) kind-of confirms that. I think there is huge benefit with standardization so btree make sense.

At the high-level, what we really need is some data-structure with an:
- efficient copy-on-write semantics 
- efficient access to individual items and traversal by a predefined order
I wouldn't exclude that there may be other structures supporting that, but I consider BTree a standard solution for that and the fact that we can use an implementation that is used in other projects (including etcd itself) is a huge benefit imho.

You request to define all of that in more detail (I believe we should use this to also precisely define what semantic we expect from the underlying storage) is a valid one - and I'm happy to take an action item for that. But I don't think we should block this KEP on that.

Also - one question - do you have any preference about what format this should be written in? I don't think it should be a KEP, but maybe there is some good format in which this can be written.

@smarterclayton ^^


> This is a good point but it makes me wonder if we'd just be better off moving etcd completely in-tree and running it directly out of the apiserver process.

@logicalhan - I'm really against linking etcd into kube-apiserver for two main reasons:
- not everyone needs to back their cluster with etcd - https://github.com/k3s-io/kine is a good example of project to look at
- even with etcd, I really want to be able to scale number of kube-apiservers independently from the number of etcd instances underneath